### PR TITLE
Use `aws-lambda-java-serialization` library, which is available by default, while deserializing input and serializing output

### PR DESCRIPTION
**Pull request by serkan-ozal**
Fixes https://github.com/open-telemetry/opentelemetry-lambda/issues/640

## The Problem
AWS Lambda Java Runtime has its own customized serialization mechanism (mixin, etc ...). So to be able to deserialize input and serialize output properly, we need to delegate their serializers/deserializers. Fortunately, their serialization/deserialization implementation is published in the `aws-lambda-java-serialization` library.

## Highlights of the PR 
- `TracingRequestWrapper` becomes a `RequestStreamHandler` implementation (it was a `RequestHandler` implementation before)
-  In the `TracingRequestStreamHandler`, the root/main span of the handler is no longer ended on the `OutputStream` close, but at the end of the user handler execution.
- On the `OutputStream` close in the `TracingRequestStreamHandler`, an event (mentioning about the `OutputStream` has been closed) is added to the root/main span of the handler.

## The Execution Flow
- Initially, `TracingRequestWrapper` starts the root/main span of the handler 
- Then, `TracingRequestWrapper` gets input as `InputStream` and deserializes it to the target object type based on the input type of the wrapped user handler by using the deserializers in the `aws-lambda-java-serialization` library.
- Next, `TracingRequestWrapper` calls the wrapped user handler with the deserialized input, waits it to be completed and gets its output (response).
- Later on, `TracingRequestWrapper` serializes the output into the `OutputStream` (so it will be sent back to the client by AWS Lambda runtime after all the Lambda invocation/execution has completed)
- And in the end, `TracingRequestWrapper` ends the root/main span of the handler and flushes the created spans throughout the exporter.

### CODE REVIEW

**Comment by serkan-ozal** [comment]
@tylerbenson @rapphil Ready for review

**Comment by serkan-ozal** [comment]
After this PR, I think, we can also unify handlers for all kinds of events by deprecating individual SQS and API GW handlers.

**Comment by rapphil** [comment]
Thanks for your PR! I will need a bit of time to understand all the changes.

**Comment by rapphil** [comment]
Can you add a PR description describing your approach? Thank you for taking the time to describe your changes in the WG meeting today.

**Comment by serkan-ozal** [comment]
@tylerbenson @rapphil could you please check my latest commit whether it applies the changes we agreed on in the SIG meeting.

**Comment by serkan-ozal** [comment]
@rapphil @tylerbenson  I have updated the description, could you please check that

**Review by tylerbenson — APPROVED** [review]
One suggestion, one minor question... otherwise looks good to me.

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/build.gradle.kts [hunk 1]
```diff
@@ -18,6 +18,8 @@ dependencies {
   // in public API.
   library("com.amazonaws:aws-lambda-java-events:2.2.1")
 
+  compileOnly("com.amazonaws:aws-lambda-java-serialization:1.1.5")
```
Comment (rapphil) [inline instrumentation/aws-lambda/aws-lambda-events-2.2/library/build.gradle.kts:21]: Maybe add a comment here about this library being available during runtime natively in lambda?

Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/build.gradle.kts:26]: @rapphil done

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java [hunk 2]
```diff
@@ -5,32 +5,90 @@
...
 
 /**
- * Wrapper for {@link io.opentelemetry.instrumentation.awslambdacore.v1_0.TracingRequestHandler}.
- * Allows for wrapping a regular lambda, not proxied through API Gateway. Therefore, HTTP headers
- * propagation is not supported.
+ * Wrapper for {@link com.amazonaws.services.lambda.runtime.RequestHandler} based Lambda handlers.
  */
-public class TracingRequestWrapper extends TracingRequestWrapperBase<Object, Object> {
+public class TracingRequestWrapper extends TracingRequestStreamWrapper {
   public TracingRequestWrapper() {
```
[the remark this answers is not in the document]
Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java:31]: @tylerbenson default constructors are used by the AWS Lambda runtime while instantiating our wrapper handlers. Other constructors with arguments are used by the tests.

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public final class SerializationUtil {
+
+  private static final Map<Class<?>, PojoSerializer<?>> serializerCache = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java]: @tylerbenson done

**Review by tylerbenson — APPROVED** [review]

**Review by rapphil — APPROVED** [review]

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      try {
+        objectMapper.writeValue(output, value);
+      } catch (IOException e) {
+        throw new IllegalStateException("Could not serialize to JSON.", e);
+      }
+    }
+  }
+
+  private SerializationUtil() {}
+}
```
Comment (breedx-splk) [inline instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java:133]: I'd like to see new classes like this show up with unit test coverage.

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/test/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapperStandardEventsTest.java [hunk 5]
```diff
@@ -15,198 +15,215 @@
...
+    final String eventBody;
+
+    EventInfo(Class<?> eventType, String eventBody) {
+      this.eventType = eventType;
+      this.eventBody = eventBody;
+    }
+  }
+
+  private static Map<Class<?>, EventInfo> buildEventExamples() {
+    Map<Class<?>, EventInfo> events = new HashMap<>();
```
Comment (breedx-splk) [inline instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/test/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapperStandardEventsTest.java:45]: Probably want to use `ClassValue` here as well for consistency and the reasons @tylerbenson mentioned previously.

Comment (tylerbenson) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/test/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapperStandardEventsTest.java:45]: I'm less concerned about it in a test class, but 🤷.

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java [hunk 6]
```diff
@@ -5,32 +5,90 @@
...
+  // Used for testing
+  <I, O> O handleRequest(I input, Context context) throws IOException {
+    byte[] inputJsonData = SerializationUtil.toJsonData(input);
+    ByteArrayInputStream inputStream = new ByteArrayInputStream(inputJsonData);
+    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+    super.handleRequest(inputStream, outputStream, context);
+
+    byte[] outputJsonData = outputStream.toByteArray();
+    return (O)
```
Comment (breedx-splk) [inline instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java:82]: Something visually strange about this to me. :) Can we just call these types `INPUT` and `OUTPUT` for clarity then?

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java [hunk 7]
```diff
@@ -5,32 +5,90 @@
...
+    Object[] parameters = LambdaParameters.toParameters(targetMethod, request.getInput(), context);
+    try {
+      Object result = targetMethod.invoke(wrappedLambda.getTargetObject(), parameters);
+      SerializationUtil.toJson(output, result);
+    } catch (IllegalAccessException e) {
+      throw new IllegalStateException("Method is inaccessible", e);
+    } catch (InvocationTargetException e) {
+      throw (e.getCause() instanceof RuntimeException
+          ? (RuntimeException) e.getCause()
+          : new IllegalStateException(e.getTargetException()));
```
Comment (breedx-splk) [inline instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java:68]: Why are we stripping off the top layer of the exception chain? Is it to remove the instrumentation itself or something?

**Review by breedx-splk — COMMENTED** [review]
Thanks for the contribution and sorry for the delay. Had a few comments, but I think this is definitely on the right track. Appreciate it!

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java (hunk 7 above)
Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java:68]: It is the same exception handling logic borrowed from `TracingRequestWrapperBase.doHandleRequest` which was parent class of `TracingRequestWrapper` before these changes. Maybe @tylerbenson or @rapphil can enlighten us on the reason behind this logic.

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java (hunk 4 above)
Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java:133]: Yep, this class is indirectly tested by existing and new tests but not directly. I will add tests to test `SerializationUtil` directly.

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/test/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapperStandardEventsTest.java (hunk 5 above)
Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/test/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapperStandardEventsTest.java:45]: I couldn't get it how to use `ClassValue` here. There is no lazy retrieved value here and not sure `ClassValue` fits better here than `Map`

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java (hunk 6 above)
Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java]: done

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java (hunk 4 above)
Comment (serkan-ozal) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/internal/SerializationUtil.java:133]: done

**Comment by serkan-ozal** [comment]
Hi @breedx-splk, Thanks for the review. Could you please check my responses and commits.

File: instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/test/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapperStandardEventsTest.java (hunk 5 above)
Comment (breedx-splk) [reply instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/test/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapperStandardEventsTest.java:45]: Ah yeah ok test class, it's probably fine.

**Review by breedx-splk — APPROVED** [review]
Thanks for the contributions!

**Pull request merged**